### DISCUSSION TITLE
Failure 'Check Tvheadend[class=XPathActivity] failed'

### DISCUSSION THREAD

**Discussion by malvinas2** [languitar/autosuspend#483] [General]
After I have been using the program on a computer running Ubuntu 16.04 for several years, I have now also installed it on a second computer running Ubuntu 22.04. Since I installed it from the package sources, it is probably still version 4.0.1.
The configuration file is identical to that I used within 16.04, I only adjusted the paths where necessary.
However, I get the error message that the TVHeadend check failed, see log. TVHeadend is still running on the old computer, but it can be accessed at the URL specified in the configuration file, and the username and password are also correct. 

Does anyone have an idea what the error might be?

EDIT: I also installed TVHeadend on the new computer and modified the config of autosuspend, directing the ip-adress to the local one, but the error remains. 

[autosuspend.conf.txt](https://github.com/languitar/autosuspend/files/14132013/autosuspend.conf.txt)
[log-autosuspend.log.txt](https://github.com/languitar/autosuspend/files/14132016/log-autosuspend.log.txt)

**Comment by languitar** [comment]
That exception pretty clearly says that autosuspend on the system is not able to open a connection to host='192.168.1.14', port=9981. From the system, can you use curl to do something like `curl -v http://192.168.1.14:9981/status.xml`?

**Comment by malvinas2** [comment]
`curl `show no problems. 

And when I change the URL from the local home network IP of the new computer (192.168.1.41) to 127.0.0.1 it works, see log file. 
(the old computer has got 192.168.1.14, the new one 192.168.1.41 :-)  )

[autosuspend_log_localhost.txt](https://github.com/languitar/autosuspend/files/14171558/autosuspend_log_localhost.txt)
[curl_192_168_1_41.txt](https://github.com/languitar/autosuspend/files/14171561/curl_192_168_1_41.txt)

**Reply by languitar** [reply]
The answer looks pretty simple now that I had a closer look at your config: 14 is not 41. The last segment of the IP is different. If you curl for one IP and autosuspend uses a different one, then at least I can't see if the configured IP is working as expected with curl.

**Reply by malvinas2** [reply]
No, it is simpler: 
Old computer (ubuntu 16.04): 192.168.1.**14**  --> autosuspend works with `url = http://192.168.1.14:9981/status.xml`

New computer (ubuntu 22.04): 192.168.1.**41** --> autosuspend does not work with `url = http://192.168.1.41:9981/status.xml` neither does it connect to the other computer `url = http://192.168.1.14:9981/status.xml` (this was just for testing purposes as it obviously has limited use), but `url = http://127.0.0.1:9981/status.xml` does the job. 

(sorry für die Verwirrung....)

**Reply by languitar** [reply]
So maybe the tvheadend installation is listening differently on both computers. `ss -tlnp` shows which process is listening on which ports and interfaces. That output might help.

**Reply by malvinas2** [reply]
Yes, that's probably it: 

```
> ss -tlnp
State	Recv-Q	Send-Q		Local Address:Port	Peer Address:Port
LISTEN	0	127			*:9981			*:*		<-- OLD computer (.14)
users:(("tvheadend",pid=25305,fd=6))

LISTEN	0	511		0.0.0.0:9981		0.0.0.0:*		<-- NEW computer (.41)
users:(("tvheadend",pid=1414,fd=8))
```
According to stackexchange: 
```
0.0.0.0:22	listening on all IPv4 Interfaces only
[::]:22		IPv6 only
*:22		IPv4 and IPv6
```
~~Well, the whole thing obviously has nothing to do with Autosuspend and you are not the 'Auskunft', 
but I would have guessed that "IPv4 and IPv6" is better than just "IPv4". 
If I want to change this on the new computer, what keyword should I google?
(a firewall is not active or even installed on any of the computers)~~


EDIT: Getting back to Autosuspend:
From my understanding `*:9981` or `0.0.0.0:9981` shouldn't matter. If the entry on the new computer was `127.0.0.1:9981`, I would understand it more easily.

**Reply by languitar** [reply]
Doesn't look like the reason. All installations listen at least on all IPv4 addresses and interfaces. So you should be able to connect when using an IPv4 address.

**Reply by malvinas2** [reply]
Well, with 127.0.0.1 as URL it works now.
I don't know how useful it is to continue searching for the cause. Especially since no other user has reported problems with this so far. Therefore, I would close this discussion for now.

**Reply by languitar** [reply]
Would be interesting to know what's causing this in case you still find out.

**Comment by malvinas2** [comment]
The problem is back, now with a slightly different error message: 
`Read timed out. (read timeout=5)` instead of `Max retries` (also in `Check Tvheadend[class=XPathActivity]` ). 

First, I verified having used '`http`' and not '`httpS`' in the config-file, which was true. 
Secondly, I didn't reboot the computer this time, but restarted the systemd-unit: `systemctl restart autosuspend.service`. And this worked! 

So, I imagine the check for TVHeadend fails because TVH hasn't started completely. 
I added an '`ExecStartPre=/bin/sleep 5`' command in the systemd configuration file of autosuspend and this seems to do the job. 
(creating a target and using before/after would be more elegant, but well... ExecStartPre works, too). 

[log-autosuspend.log_2.txt](https://github.com/languitar/autosuspend/files/14462857/log-autosuspend.log_2.txt)
[autosuspend.service.txt](https://github.com/languitar/autosuspend/files/14462864/autosuspend.service.txt)

**Reply by languitar** [reply]
The exception should only be logged in that iteration. Once tvheadend becomes available, the next checking iteration will work. I cannot provide any dependencies on the systemd services as I don't know who is using which service.
